File: Data_Structures/nealwu_dsu.cpp
#include <bits/stdc++.h>

using namespace std;

// efficient & minimized DSU implementation stolen from Neal Wu
struct DSU {
    vector<int> data;
    vector<int> min_el;
    vector<int> max_el;

    // usually want to pass n+1 to constructor
    explicit DSU(int n) {
        data.assign(n, -1);
        min_el.resize(n);
        max_el.resize(n);
        for (int i = 0; i < n; i++) {
            min_el[i] = max_el[i] = i;
        }
    }

    int find(int x) {
        return data[x] < 0 ? x : data[x] = find(data[x]);
    }

    void union_sets(int x, int y) {
        x = find(x);
        y = find(y);
        if (x == y) return;

        if (data[x] > data[y]) swap(x, y);

        data[x] += data[y];
        data[y] = x;

        min_el[x] = min(min_el[x], min_el[y]);
        max_el[x] = max(max_el[x], max_el[y]);
    }

    int set_size(int x) {
        return (-data[find(x)]);
    }
};

File: datastructures/gol/dsu/nealwu_dsu.go
package dsu

// efficient & minimized DSU implementation stolen from Neal Wu
// don't know who came up with this, but super cute
type NealWuDSU struct {
	data  []int
	minEl []int
	maxEl []int
}

// usually want to pass n+1 to constructor
func NewNealWuDSU(n int) *NealWuDSU {
	d := &NealWuDSU{
		data:  make([]int, n),
		minEl: make([]int, n),
		maxEl: make([]int, n),
	}
	for i := 0; i < n; i++ {
		d.data[i] = -1
		d.minEl[i] = i
		d.maxEl[i] = i
	}
	return d
}

func (d *NealWuDSU) Find(x int) int {
	if d.data[x] < 0 {
		return x
	}
	d.data[x] = d.Find(d.data[x])
	return d.data[x]
}

func (d *NealWuDSU) UnionSets(x, y int) {
	x = d.Find(x)
	y = d.Find(y)
	if x == y {
		return
	}

	if d.data[x] > d.data[y] {
		x, y = y, x
	}

	d.data[x] += d.data[y]
	d.data[y] = x

	if d.minEl[y] < d.minEl[x] {
		d.minEl[x] = d.minEl[y]
	}
	if d.maxEl[y] > d.maxEl[x] {
		d.maxEl[x] = d.maxEl[y]
	}
}

func (d *NealWuDSU) SetSize(x int) int {
	return -d.data[d.Find(x)]
}
